### DISCUSSION TITLE
Best practices for workspace structure in terms of scalability, maintainability and caching

### DISCUSSION THREAD

**Discussion by joafeldmann** [vercel/turborepo#5573] [Help] [answered]
In a scenario where most of the code can be shared between two apps and packages are not intended to be published, what would be the best workspace structure in terms of scalability, maintainability and caching?

Having a single shared package:

```
apps
  appA
  appB
packages
  shared
    components
    feature-A
    feature-B
    feature-C
    ....
    feature-XYZ
    hooks
    services
    utils
    package.json
```

Having multiple shared packages:

```
apps
  appA
  appB
packages
  shared
    components
      package.json
    feature-A
      package.json
    feature-B
      package.json
    feature-C
      package.json
    ....
    feature-XYZ
      package.json
    hooks
      package.json
    services
      package.json
    utils
      package.json
```

**Comment by anthonyshew** [comment]
There is no one, clean answer we can provide for this question. Your business requirements, tools you're building with, scale, personnel, and many other factors should dictate how you architect your monorepo and drive the design that you create.

What you've describe is a mix of domain- and utility-based packaging strategies, a common way of creating a monorepo. But, like I said before, this question has a lot of inputs to get to the right answer so we likely won't be able to get you a great answer unless you're specific about your design needs.

**Reply by anthonyshew** [reply, marked as answer]
Gave a shot at explaining this here: https://turbo.build/repo/docs/crafting-your-repository/creating-an-internal-package#one-purpose-per-package
